curvlinops/ekfac.py

```diff
@@ -0,0 +1,638 @@
+        Will call ``compute_kronecker_factors`` and ``compute_eigenvalue_correction`` if
+        either of them has not been called before and will cache the logdet until one of
+        them is called again.
+
+        Returns:
+            Log determinant of the EKFAC approximation.
+        """
+        if self._logdet is not None:
+            return self._logdet
+
+        self._maybe_compute_ekfac()
+
+        # Compute the log determinant using the corrected eigenvalues
+        self._logdet = 0.0
+        for corrected_eigenvalues in self._corrected_eigenvalues.values():
+            if isinstance(corrected_eigenvalues, dict):
+                for val in corrected_eigenvalues.values():
+                    self._logdet += val.log().sum()
+            else:
+                self._logdet += corrected_eigenvalues.log().sum()
+
+        return self._logdet
+
+    @property
+    def frobenius_norm(self) -> Tensor:
+        r"""Frobenius norm of the EKFAC approximation.
+
+        Will call ``compute_kronecker_factors`` and ``compute_eigenvalue_correction`` if
+        either of them has not been called before and will cache the Frobenius norm
+        until one of them is called again.
+
+        Returns:
+            Frobenius norm of the EKFAC approximation.
+        """
+        if self._frobenius_norm is not None:
+            return self._frobenius_norm
+
+        self._maybe_compute_ekfac()
+
+        # Compute the Frobenius norm using the corrected eigenvalues
+        self._frobenius_norm = 0.0
+        for corrected_eigenvalues in self._corrected_eigenvalues.values():
+            if isinstance(corrected_eigenvalues, dict):
+                for val in corrected_eigenvalues.values():
+                    self._frobenius_norm += val.square().sum()
+            else:
+                self._frobenius_norm += corrected_eigenvalues.square().sum()
+
+        return self._frobenius_norm.sqrt_()
+
+    def state_dict(self) -> Dict[str, Any]:
+        """Return the state of the EKFAC linear operator.
+
+        Returns:
+            State dictionary.
+        """
+        state_dict = super().state_dict()
+        # Add quantities specifically for EKFAC (if computed)
+        state_dict.update(
+            {
+                "input_covariances_eigenvectors": self._input_covariances_eigenvectors,
+                "gradient_covariances_eigenvectors": self._gradient_covariances_eigenvectors,
+                "cached_activations": self._cached_activations,
+                "corrected_eigenvalues": self._corrected_eigenvalues,
+            }
+        )
+        return state_dict
+
+    def load_state_dict(self, state_dict: Dict[str, Any]):
+        """Load the state of the EKFAC linear operator.
+
+        Args:
+            state_dict: State dictionary.
+        """
+        super().load_state_dict(state_dict)
+
+        # Set EKFAC-specific quantities
+        # TODO: should we check if the keys match the mapping keys?
```

Comment:
I think this would be good because this will allow detecting inconsistencies when loading a state dict that would otherwise lead to errors only much later (e.g. in a matvec).